src/moves.rs

```diff
@@ -352,87 +337,117 @@ where
     GetStackSlot: FnMut() -> Allocation,
     IsStackAlloc: Fn(Allocation) -> bool,
 {
-    pub fn new(
-        find_free_reg: GetReg,
-        get_stackslot: GetStackSlot,
-        is_stack_alloc: IsStackAlloc,
-        victim: PReg,
-    ) -> Self {
-        Self {
-            stack_stack_scratch_reg: None,
-            stack_stack_scratch_reg_save: None,
-            find_free_reg,
-            get_stackslot,
-            is_stack_alloc,
-            victim,
+    pub fn compute<T: Debug + Default + Copy>(
+        mut self,
+        moves: MoveVecWithScratch<T>,
+    ) -> MoveVec<T> {
+        let moves = if moves.needs_scratch() {
+            // Now, find a scratch allocation in order to resolve cycles.
+            let scratch = (self.find_free_reg)().unwrap_or_else(|| (self.get_stackslot)());
+            trace!("scratch resolver: scratch alloc {:?}", scratch);
+
+            moves.with_scratch(scratch)
+        } else {
+            moves.without_scratch().unwrap()
+        };
+
+        // Do we have any stack-to-stack moves? Fast return if not.
+        let stack_to_stack = moves
+            .iter()
+            .any(|&(src, dst, _)| (self.is_stack_alloc)(src) && (self.is_stack_alloc)(dst));
+        if !stack_to_stack {
+            return moves;
         }
-    }
 
-    pub fn compute<T: Debug + Copy>(mut self, moves: MoveVecWithScratch<T>) -> MoveVec<T> {
-        // First, do we have a vec with no stack-to-stack moves or use
-        // of a scratch register? Fast return if so.
-        if !moves.needs_scratch() && !moves.stack_to_stack(&self.is_stack_alloc) {
-            return moves.without_scratch().unwrap();
-        }
+        // Allocate a scratch register for stack-to-stack move expansion.
+        let (scratch_reg, save_slot) = if let Some(reg) = (self.find_free_reg)() {
+            trace!(
+                "scratch resolver: have free stack-to-stack scratch preg: {:?}",
+                reg
+            );
+            (reg, None)
+        } else {
+            let reg = Allocation::reg(self.victim);
+            // Stackslot into which we need to save the stack-to-stack
+            // scratch reg before doing any stack-to-stack moves, if we stole
+            // the reg.
+            let save = (self.get_stackslot)();
+            trace!(
+                "scratch resolver: stack-to-stack using victim {:?} with save stackslot {:?}",
+                reg,
+                save
+            );
+            (reg, Some(save))
+        };
+
+        // Mutually exclusive flags for whether either scratch_reg or
+        // save_slot need to be restored from the other. Initially,
+        // scratch_reg has a value we should preserve and save_slot
+        // has garbage.
+        let mut scratch_dirty = false;
+        let mut save_dirty = true;
```

Comment:
Could we define an enum to track the state of the two locations and/or the saveslot? I think the logic here is correct but something is unsatisfying and slightly hard to follow/verify about the ad-hoc ifs -- it feels like maybe it should be a state machine.